File: app/src/main/java/com/example/ui/CurioComponents.kt
package com.example.ui

import android.graphics.Bitmap
import android.graphics.ImageDecoder
import android.net.Uri
import android.os.Build
import android.provider.MediaStore
import androidx.activity.compose.rememberLauncherForActivityResult
import androidx.activity.result.contract.ActivityResultContracts
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.ExperimentalLayoutApi
import androidx.compose.foundation.layout.FlowRow
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxHeight
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.border
import androidx.compose.foundation.horizontalScroll
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Autorenew
import androidx.compose.material.icons.filled.BarChart
import androidx.compose.material.icons.filled.Bookmarks
import androidx.compose.material.icons.filled.Close
import androidx.compose.material.icons.filled.CloudSync
import androidx.compose.material.icons.filled.ContentCopy
import androidx.compose.material.icons.filled.DeleteSweep
import androidx.compose.material.icons.filled.Psychology
import androidx.compose.material.icons.filled.Refresh
import androidx.compose.material.icons.filled.Screenshot
import androidx.compose.material.icons.filled.Link
import androidx.compose.material.icons.filled.Search
import androidx.compose.material.icons.filled.Settings
import androidx.compose.material.icons.filled.Share
import androidx.compose.material.icons.filled.Timer
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.text.style.TextDecoration
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.window.Dialog
import androidx.compose.foundation.ExperimentalFoundationApi
import androidx.compose.foundation.combinedClickable
import androidx.compose.material.icons.filled.Add
import androidx.compose.material.icons.filled.Check
import androidx.compose.material.icons.filled.Delete
import androidx.compose.material.icons.filled.Label
import androidx.compose.material.icons.filled.Menu
import androidx.compose.material.icons.filled.ArrowBack
import androidx.compose.material.icons.filled.KeyboardArrowUp
import androidx.compose.material.icons.filled.KeyboardArrowDown
import androidx.compose.material.icons.outlined.BarChart
import androidx.compose.material.icons.outlined.Bookmarks
import androidx.compose.material.icons.outlined.Settings
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.IconButton
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.RadioButton
import androidx.compose.material3.Switch
import androidx.compose.material3.Text
import androidx.compose.ui.draw.clip
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.compose.foundation.layout.widthIn
import androidx.compose.foundation.layout.safeDrawingPadding
import androidx.compose.foundation.layout.imePadding
import androidx.compose.foundation.layout.statusBarsPadding
import androidx.compose.foundation.layout.WindowInsets
import androidx.compose.foundation.layout.isImeVisible
import androidx.compose.foundation.verticalScroll
import androidx.compose.foundation.isSystemInDarkTheme
import androidx.compose.animation.AnimatedVisibility
import androidx.compose.animation.expandVertically
import androidx.compose.animation.shrinkVertically
import androidx.compose.animation.fadeIn
import androidx.compose.animation.fadeOut
import androidx.compose.material.icons.filled.ExpandLess
import androidx.compose.material.icons.filled.ExpandMore
import androidx.compose.material.icons.filled.AutoAwesome
import android.content.Intent
import kotlinx.coroutines.launch
import com.example.MainActivity
import com.example.domain.model.AuthState
import com.example.domain.model.Bookmark
import com.example.domain.model.SourceType
import com.example.ui.components.GlassBottomBar
import com.example.ui.components.GlassNavigationItem
import com.example.ui.components.GlassScaffold
import com.example.ui.components.GlassTopBar
import com.example.ui.components.LiquidGlassFab
import com.example.ui.screens.auth.AuthViewModel
import com.example.ui.screens.auth.LoginScreen
import com.example.ui.theme.BookmarkTheme
import com.example.ui.theme.GlassTier
import com.example.ui.theme.glassSurface
import com.example.ui.theme.rememberGlassTier
import com.example.ui.theme.CurioMotion
import com.example.ui.theme.pressBounce
import com.example.ui.theme.bounceScale
import com.example.ui.theme.curioAccentBrush
import coil.compose.AsyncImage
import coil.compose.SubcomposeAsyncImage
import androidx.compose.foundation.layout.aspectRatio
import androidx.compose.foundation.layout.heightIn
import androidx.compose.foundation.layout.offset
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.draw.rotate
import androidx.compose.ui.draw.scale
import androidx.compose.ui.draw.alpha
import androidx.compose.animation.core.animateFloatAsState
import androidx.compose.animation.core.animateFloat
import androidx.compose.animation.animateContentSize
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.animation.core.rememberInfiniteTransition
import androidx.compose.animation.core.infiniteRepeatable
import androidx.compose.animation.core.RepeatMode
import androidx.compose.animation.core.tween
import androidx.compose.foundation.interaction.MutableInteractionSource
import androidx.compose.foundation.interaction.collectIsPressedAsState
import androidx.compose.foundation.lazy.rememberLazyListState
import androidx.compose.material.icons.filled.MoreHoriz
import androidx.compose.material.icons.automirrored.filled.Send
import androidx.compose.material.icons.filled.Article
import androidx.compose.material.icons.filled.TrendingUp
import androidx.compose.material.icons.filled.LocalFireDepartment
import androidx.compose.material.icons.filled.Hub
import androidx.compose.material.icons.automirrored.filled.MenuBook
import java.text.SimpleDateFormat
import java.util.Locale

@Composable
internal fun CategoryPill(category: String) {
    val color = getCategoryColor(category)
    Box(
        modifier = Modifier
            .background(color.copy(alpha = 0.16f), RoundedCornerShape(50))
            .border(1.dp, color.copy(alpha = 0.45f), RoundedCornerShape(50))
            .padding(horizontal = 10.dp, vertical = 4.dp)
    ) {
        Text(
            text = category.uppercase(),
            style = MaterialTheme.typography.labelSmall.copy(
                fontWeight = FontWeight.Black, color = color, fontSize = 10.sp, letterSpacing = 0.8.sp
            )
        )
    }
}

@Composable
internal fun TagChip(tag: String, onClick: () -> Unit) {
    Box(
        modifier = Modifier
            .clip(RoundedCornerShape(50))
            .background(MaterialTheme.colorScheme.primary.copy(alpha = 0.10f))
            .pressBounce(onClick = onClick)
            .padding(horizontal = 10.dp, vertical = 5.dp)
    ) {
        Text(
            text = "#$tag",
            style = MaterialTheme.typography.labelSmall.copy(
                fontWeight = FontWeight.Bold, color = MaterialTheme.colorScheme.primary
            )
        )
    }
}

@Composable
internal fun CurioActionChip(label: String, icon: androidx.compose.ui.graphics.vector.ImageVector, color: Color, filled: Boolean = false, onClick: () -> Unit) {
    Row(
        modifier = Modifier
            .clip(RoundedCornerShape(12.dp))
            .background(if (filled) color else color.copy(alpha = 0.12f))
            .pressBounce(onClick = onClick)
            .padding(horizontal = 12.dp, vertical = 7.dp),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(5.dp)
    ) {
        Icon(icon, contentDescription = label, modifier = Modifier.size(14.dp), tint = if (filled) MaterialTheme.colorScheme.onPrimary else color)
        Text(label, style = MaterialTheme.typography.labelSmall.copy(fontWeight = FontWeight.ExtraBold, color = if (filled) MaterialTheme.colorScheme.onPrimary else color))
    }
}

@Composable
internal fun DetailPanel(label: String, body: String, accent: Color, bold: Boolean = false) {
    Box(
        modifier = Modifier
            .fillMaxWidth()
            .background(accent.copy(alpha = 0.10f), RoundedCornerShape(12.dp))
            .border(1.dp, accent.copy(alpha = 0.20f), RoundedCornerShape(12.dp))
            .padding(12.dp)
    ) {
        Column(verticalArrangement = Arrangement.spacedBy(4.dp)) {
            Text(label, style = MaterialTheme.typography.labelSmall.copy(fontWeight = FontWeight.Black, color = accent, letterSpacing = 0.8.sp))
            Text(body, style = MaterialTheme.typography.bodySmall.copy(fontWeight = if (bold) FontWeight.Bold else FontWeight.Normal, lineHeight = 19.sp), color = MaterialTheme.colorScheme.onSurface.copy(alpha = 0.9f))
        }
    }
}

/** Detail panel whose body is rendered as markdown (headings, bullets, **bold**). */
@Composable
internal fun MarkdownDetailPanel(label: String, body: String, accent: Color, icon: androidx.compose.ui.graphics.vector.ImageVector? = null) {
    Box(
        modifier = Modifier
            .fillMaxWidth()
            .background(accent.copy(alpha = 0.10f), RoundedCornerShape(12.dp))
            .border(1.dp, accent.copy(alpha = 0.20f), RoundedCornerShape(12.dp))
            .padding(12.dp)
    ) {
        Column(verticalArrangement = Arrangement.spacedBy(8.dp)) {
            Row(verticalAlignment = Alignment.CenterVertically, horizontalArrangement = Arrangement.spacedBy(5.dp)) {
                if (icon != null) Icon(icon, contentDescription = null, tint = accent, modifier = Modifier.size(13.dp))
                Text(label, style = MaterialTheme.typography.labelSmall.copy(fontWeight = FontWeight.Black, color = accent, letterSpacing = 0.8.sp))
            }
            MarkdownText(
                markdown = body,
                style = MaterialTheme.typography.bodySmall.copy(lineHeight = 19.sp),
                color = MaterialTheme.colorScheme.onSurface.copy(alpha = 0.92f),
                accent = accent
            )
        }
    }
}

@Composable
internal fun StatTile(label: String, value: String, icon: androidx.compose.ui.graphics.vector.ImageVector, color: Color, tier: GlassTier, modifier: Modifier = Modifier) {
    Box(modifier = modifier.glassSurface(tier = tier, shape = RoundedCornerShape(18.dp)).padding(14.dp)) {
        Column(verticalArrangement = Arrangement.spacedBy(6.dp)) {
            Box(modifier = Modifier.size(30.dp).clip(RoundedCornerShape(10.dp)).background(color.copy(alpha = 0.16f)), contentAlignment = Alignment.Center) {
                Icon(icon, contentDescription = null, tint = color, modifier = Modifier.size(17.dp))
            }
            Text(value, style = MaterialTheme.typography.headlineSmall.copy(fontWeight = FontWeight.Black), color = MaterialTheme.colorScheme.onSurface)
            Text(label.uppercase(), style = MaterialTheme.typography.labelSmall.copy(fontWeight = FontWeight.Bold, letterSpacing = 0.6.sp), color = MaterialTheme.colorScheme.onSurface.copy(alpha = 0.55f))
        }
    }
}

/**
 * Elegant gradient cover shown when a bookmark has no real or generated image.
 * Keeps the feed visually consistent — every card has a media anchor instead of
 * collapsing into a bare text block.
 */
@Composable
internal fun CurioFallbackCover(
    bookmark: Bookmark,
    srcColor: Color,
    accent: Color,
    isExpanded: Boolean
) {
    val glyph = when (bookmark.sourceType) {
        SourceType.GITHUB -> Icons.Filled.Hub
        SourceType.ARXIV -> Icons.AutoMirrored.Filled.MenuBook
        SourceType.HUGGING_FACE -> Icons.Filled.AutoAwesome
        else -> Icons.Filled.Bookmarks
    }
    Box(
        modifier = Modifier
            .fillMaxWidth()
            .height(if (isExpanded) 104.dp else 70.dp)
            .clip(RoundedCornerShape(16.dp))
            .background(Brush.linearGradient(listOf(srcColor.copy(alpha = 0.9f), accent.copy(alpha = 0.45f))))
            .border(1.dp, Color.White.copy(alpha = 0.12f), RoundedCornerShape(16.dp))
    ) {
        // Oversized translucent watermark glyph
        Icon(
            glyph,
            contentDescription = null,
            tint = Color.White.copy(alpha = 0.22f),
            modifier = Modifier
                .align(Alignment.CenterEnd)
                .padding(end = 6.dp)
                .size(if (isExpanded) 92.dp else 64.dp)
        )
        Column(
            modifier = Modifier.align(Alignment.BottomStart).padding(14.dp),
            verticalArrangement = Arrangement.spacedBy(2.dp)
        ) {
            Text(
                text = sourceDisplayName(bookmark).uppercase(),
                style = MaterialTheme.typography.labelMedium.copy(fontWeight = FontWeight.Black, letterSpacing = 1.sp),
                color = Color.White
            )
            if (!bookmark.category.isNullOrBlank()) {
                Text(
                    text = "#${bookmark.category.lowercase().replace(' ', '_')}",
                    style = MaterialTheme.typography.labelSmall.copy(fontWeight = FontWeight.Bold),
                    color = Color.White.copy(alpha = 0.85f)
                )
            }
        }
    }
}

/** Compact circular glass icon button used in the feed control strip. */
@Composable
internal fun FeedIconAction(
    icon: androidx.compose.ui.graphics.vector.ImageVector,
    contentDescription: String,
    active: Boolean = false,
    spinning: Boolean = false,
    onClick: () -> Unit
) {
    val tint = if (active) MaterialTheme.colorScheme.primary else MaterialTheme.colorScheme.onSurface.copy(alpha = 0.7f)
    val spin = rememberInfiniteTransition(label = "spin")
    val angle by spin.animateFloat(
        initialValue = 0f, targetValue = 360f,
        animationSpec = infiniteRepeatable(tween(900, easing = androidx.compose.animation.core.LinearEasing)),
        label = "spinAngle"
    )
    Box(
        modifier = Modifier
            .size(38.dp)
            .clip(CircleShape)
            .background(if (active) MaterialTheme.colorScheme.primary.copy(alpha = 0.16f) else MaterialTheme.colorScheme.onSurface.copy(alpha = 0.06f))
            .pressBounce(pressedScale = 0.85f, onClick = onClick),
        contentAlignment = Alignment.Center
    ) {
        Icon(
            icon,
            contentDescription = contentDescription,
            tint = tint,
            modifier = Modifier
                .size(18.dp)
                .then(if (spinning) Modifier.rotate(angle) else Modifier)
        )
    }
}

/** Quick filter pill (All / Favorites / Read later) with an optional count badge. */
@Composable
internal fun QuickFilterPill(
    label: String,
    icon: androidx.compose.ui.graphics.vector.ImageVector,
    count: Int?,
    active: Boolean,
    color: Color,
    onClick: () -> Unit
) {
    Row(
        modifier = Modifier
            .bounceScale(active)
            .clip(RoundedCornerShape(50))
            .background(if (active) color.copy(alpha = 0.18f) else MaterialTheme.colorScheme.onSurface.copy(alpha = 0.05f))
            .border(1.dp, if (active) color.copy(alpha = 0.55f) else Color.Transparent, RoundedCornerShape(50))
            .pressBounce(pressedScale = 0.93f, onClick = onClick)
            .padding(horizontal = 12.dp, vertical = 7.dp),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(5.dp)
    ) {
        Icon(icon, contentDescription = null, modifier = Modifier.size(14.dp), tint = if (active) color else MaterialTheme.colorScheme.onSurface.copy(alpha = 0.6f))
        Text(
            label,
            style = MaterialTheme.typography.labelMedium.copy(fontWeight = FontWeight.Bold),
            color = if (active) color else MaterialTheme.colorScheme.onSurface.copy(alpha = 0.75f)
        )
        if (count != null && count > 0) {
            Box(
                modifier = Modifier
                    .clip(RoundedCornerShape(50))
                    .background(if (active) color else MaterialTheme.colorScheme.onSurface.copy(alpha = 0.12f))
                    .padding(horizontal = 6.dp, vertical = 1.dp)
            ) {
                Text(
                    count.toString(),
                    style = MaterialTheme.typography.labelSmall.copy(fontWeight = FontWeight.Black, fontSize = 10.sp),
                    color = if (active) Color.White else MaterialTheme.colorScheme.onSurface.copy(alpha = 0.7f)
                )
            }
        }
    }
}

@Composable
internal fun TypingDots(color: Color) {
    val transition = rememberInfiniteTransition(label = "typing")
    Row(horizontalArrangement = Arrangement.spacedBy(4.dp), verticalAlignment = Alignment.CenterVertically) {
        repeat(3) { i ->
            val scale by transition.animateFloat(
                initialValue = 0.5f,
                targetValue = 1f,
                animationSpec = infiniteRepeatable(
                    animation = tween(600, delayMillis = i * 150, easing = androidx.compose.animation.core.FastOutSlowInEasing),
                    repeatMode = RepeatMode.Reverse
                ),
                label = "dot$i"
            )
            Box(
                modifier = Modifier
                    .size(7.dp)
                    .scale(scale)
                    .background(color.copy(alpha = 0.4f + scale * 0.5f), CircleShape)
            )
        }
    }
}

File: ios/Curio/Curio/Components/CurioComponents.swift
//
//  CurioComponents.swift
//  Curio
//
//  Ports: app/src/main/java/com/example/ui/CurioComponents.kt
//         (CategoryPill, TagChip, CurioActionChip, DetailPanel, MarkdownDetailPanel, StatTile,
//          CurioFallbackCover, FeedIconAction, QuickFilterPill, TypingDots).
//
//  Reusable feed/detail widgets: chips, pills, panels, tiles, the no-image fallback cover,
//  the spinning circular feed-icon action, the quick-filter pill with count badge, and the
//  3-dot typing indicator.
//
//  CONVENTIONS §8/§9 (Components):
//   - uniform press feedback via the shared `.curioPressBounce` ButtonStyle (replaces ripple);
//   - non-hit-testing `.bounceScale(active:)` for selection/active states;
//   - centralized source-type → color/glyph (here via `sourceGlyph(_:)`, mirroring the Android
//     `when (sourceType)` blocks that recur across the card + cover);
//   - per-mode alphas/hex carried verbatim from the Compose modifiers;
//   - heavy typography (`.black`/`.heavy`) + explicit `.textCase(.uppercase)` + tracking;
//   - `category` (the AI taxonomy bucket) is shown ONLY on the fallback cover as a `#hashtag`,
//     exactly as Android did — it never surfaces as a first-class label elsewhere.
//
//  Format/lookup helpers consumed here (`CurioFormat.getCategoryColor`, `.sourceDisplayName`)
//  live in the Screens module per DESIGN §10; this widget layer treats them as the documented
//  static API surface (DESIGN "Notes for implementers").
//

import SwiftUI

// MARK: - CategoryPill

/// A pill showing the AI category with its deterministic color. Used in the bulk-category
/// dialog preview row; the color comes from the stable Java-hashCode palette (CONVENTIONS §10).
struct CategoryPill: View {
    let category: String

    var body: some View {
        let color = CurioFormat.getCategoryColor(category)
        Text(category.uppercased())
            .font(.system(size: 10, weight: .black))
            .tracking(0.8)
            .foregroundStyle(color)
            .padding(.horizontal, 10)
            .padding(.vertical, 4)
            .background(color.opacity(0.16), in: Capsule())
            .overlay { Capsule().stroke(color.opacity(0.45), lineWidth: 1) }
    }
}

// MARK: - TagChip

/// A tappable `#tag` chip with press bounce. Tapping selects the tag as a search filter.
struct TagChip: View {
    let tag: String
    let onTap: () -> Void

    @Environment(\.curioColors) private var colors

    var body: some View {
        Button(action: onTap) {
            Text("#\(tag)")
                .font(.system(size: 11, weight: .bold))
                .foregroundStyle(colors.primary)
                .padding(.horizontal, 10)
                .padding(.vertical, 5)
                .background(colors.primary.opacity(0.10), in: Capsule())
        }
        .buttonStyle(.curioPressBounce)
    }
}

// MARK: - CurioActionChip

/// A labelled, icon-led action chip; `filled` flips it to a solid accent button.
struct CurioActionChip: View {
    let label: String
    let systemImage: String
    let color: Color
    var filled: Bool = false
    let onTap: () -> Void

    @Environment(\.curioColors) private var colors

    var body: some View {
        Button(action: onTap) {
            HStack(spacing: 5) {
                Image(systemName: systemImage)
                    .font(.system(size: 14, weight: .semibold))
                    .frame(width: 14, height: 14)
                    .foregroundStyle(filled ? colors.onPrimary : color)
                Text(label)
                    .font(.system(size: 11, weight: .heavy))
                    .tracking(1.0)
                    .foregroundStyle(filled ? colors.onPrimary : color)
            }
            .padding(.horizontal, 12)
            .padding(.vertical, 7)
            .background(filled ? color : color.opacity(0.12),
                        in: RoundedRectangle(cornerRadius: 12, style: .continuous))
        }
        .buttonStyle(.curioPressBounce)
    }
}

// MARK: - DetailPanel

/// A tinted, bordered labelled text block used in the expanded card / reader for OCR,
/// summary, abstract, and note sections.
struct DetailPanel: View {
    let label: String
    let bodyText: String
    let accent: Color
    var bold: Bool = false

    init(label: String, body: String, accent: Color, bold: Bool = false) {
        self.label = label
        self.bodyText = body
        self.accent = accent
        self.bold = bold
    }

    @Environment(\.curioColors) private var colors

    var body: some View {
        VStack(alignment: .leading, spacing: 4) {
            Text(label)
                .font(.system(size: 11, weight: .black))
                .tracking(0.8)
                .foregroundStyle(accent)
            Text(self.bodyText)
                .font(.system(size: 12, weight: bold ? .bold : .regular))
                .lineSpacing(19 - 12)
                .foregroundStyle(colors.onSurface.opacity(0.9))
        }
        .frame(maxWidth: .infinity, alignment: .leading)
        .padding(12)
        .background(accent.opacity(0.10), in: RoundedRectangle(cornerRadius: 12, style: .continuous))
        .overlay { RoundedRectangle(cornerRadius: 12, style: .continuous).stroke(accent.opacity(0.20), lineWidth: 1) }
    }
}

// MARK: - MarkdownDetailPanel

/// Detail panel whose body is rendered as markdown (headings, bullets, **bold**). Used for
/// the deep-analysis section; an optional leading icon precedes the label.
struct MarkdownDetailPanel: View {
    let label: String
    let bodyText: String
    let accent: Color
    var systemImage: String? = nil

    init(label: String, body: String, accent: Color, systemImage: String? = nil) {
        self.label = label
        self.bodyText = body
        self.accent = accent
        self.systemImage = systemImage
    }

    @Environment(\.curioColors) private var colors

    var body: some View {
        VStack(alignment: .leading, spacing: 8) {
            HStack(spacing: 5) {
                if let systemImage {
                    Image(systemName: systemImage)
                        .font(.system(size: 13, weight: .semibold))
                        .frame(width: 13, height: 13)
                        .foregroundStyle(accent)
                }
                Text(label)
                    .font(.system(size: 11, weight: .black))
                    .tracking(0.8)
                    .foregroundStyle(accent)
            }
            MarkdownText(
                markdown: self.bodyText,
                style: CurioFont.bodySmallLineHeight19,
                color: colors.onSurface.opacity(0.92),
                accent: accent
            )
        }
        .frame(maxWidth: .infinity, alignment: .leading)
        .padding(12)
        .background(accent.opacity(0.10), in: RoundedRectangle(cornerRadius: 12, style: .continuous))
        .overlay { RoundedRectangle(cornerRadius: 12, style: .continuous).stroke(accent.opacity(0.20), lineWidth: 1) }
    }
}

// MARK: - StatTile

/// A glass stat card: a tinted icon chip over a big value over an ALL-CAPS label.
/// Used in the feed header strip and insights dashboard.
struct StatTile: View {
    let label: String
    let value: String
    let systemImage: String
    let color: Color
    let tier: GlassTier
    var onClick: (() -> Void)? = nil

    @Environment(\.curioColors) private var colors

    var body: some View {
        Group {
            if let onClick {
                Button(action: onClick) { tileContent }
                    .buttonStyle(.curioPressBounce)
            } else {
                tileContent
            }
        }
    }

    private var tileContent: some View {
        VStack(alignment: .leading, spacing: 6) {
            Image(systemName: systemImage)
                .font(.system(size: 17, weight: .semibold))
                .foregroundStyle(color)
                .frame(width: 30, height: 30)
                .background(color.opacity(0.16), in: RoundedRectangle(cornerRadius: 10, style: .continuous))
            Text(value)
                .font(.system(size: 24, weight: .black))
                .foregroundStyle(colors.onSurface)
            Text(label.uppercased())
                .font(.system(size: 11, weight: .bold))
                .tracking(0.6)
                .foregroundStyle(colors.onSurface.opacity(0.55))
        }
        .padding(14)
        .glassSurface(tier: tier, shape: RoundedRectangle(cornerRadius: 18, style: .continuous))
    }
}

/// Elegant gradient cover shown when a bookmark has no real or generated image. Keeps the feed
/// visually consistent — every card has a media anchor instead of collapsing into a bare text
/// block. Carries an oversized translucent source glyph watermark + source label + (optional)
/// category hashtag.
struct CurioFallbackCover: View {
    let bookmark: Bookmark
    let srcColor: Color
    let accent: Color
    let isExpanded: Bool

    var body: some View {
        ZStack {
            LinearGradient(
                colors: [srcColor.opacity(0.9), accent.opacity(0.45)],
                startPoint: .topLeading,
                endPoint: .bottomTrailing
            )

            // Oversized translucent watermark glyph, center-end.
            HStack {
                Spacer()
                Image(systemName: sourceGlyph(bookmark.sourceType))
                    .font(.system(size: isExpanded ? 92 : 64, weight: .regular))
                    .foregroundStyle(Color.white.opacity(0.22))
                    .padding(.trailing, 6)
            }

            // Source name + category hashtag, bottom-start.
            VStack(alignment: .leading, spacing: 2) {
                Spacer()
                HStack {
                    VStack(alignment: .leading, spacing: 2) {
                        Text(CurioFormat.sourceDisplayName(bookmark).uppercased())
                            .font(.system(size: 12, weight: .black))
                            .tracking(1.0)
                            .foregroundStyle(.white)
                        if let category = bookmark.category, !category.isBlank {
                            Text("#\(category.lowercased().replacingOccurrences(of: " ", with: "_"))")
                                .font(.system(size: 11, weight: .bold))
                                .foregroundStyle(Color.white.opacity(0.85))
                        }
                    }
                    Spacer()
                }
                .padding(14)
            }
        }
        .frame(height: isExpanded ? 104 : 70)
        .frame(maxWidth: .infinity)
        .clipShape(RoundedRectangle(cornerRadius: 16, style: .continuous))
        .overlay { RoundedRectangle(cornerRadius: 16, style: .continuous).stroke(Color.white.opacity(0.12), lineWidth: 1) }
    }
}

// MARK: - FeedIconAction

/// Compact circular glass icon button used in the feed control strip. `active` tints it
/// primary; `spinning` rotates the glyph continuously (sync in progress).
struct FeedIconAction: View {
    let systemImage: String
    let accessibilityLabel: String
    var active: Bool = false
    var spinning: Bool = false
    let onTap: () -> Void

    @Environment(\.curioColors) private var colors
    @Environment(\.accessibilityReduceMotion) private var reduceMotion
    @State private var angle: Double = 0

    private var tint: Color {
        active ? colors.primary : colors.onSurface.opacity(0.7)
    }

    var body: some View {
        Button(action: onTap) {
            Image(systemName: systemImage)
                .font(.system(size: 18, weight: .semibold))
                .foregroundStyle(tint)
                .rotationEffect(.degrees(spinning ? angle : 0))
                .frame(width: 38, height: 38)
                .background(
                    (active ? colors.primary.opacity(0.16) : colors.onSurface.opacity(0.06)),
                    in: Circle()
                )
        }
        .buttonStyle(.curioPressBounce(pressedScale: 0.85))
        .accessibilityLabel(accessibilityLabel)
        .onChange(of: spinning) { _, isSpinning in
            if isSpinning && !reduceMotion {
                angle = 0
                withAnimation(.linear(duration: 0.9).repeatForever(autoreverses: false)) {
                    angle = 360
                }
            } else {
                withAnimation(.linear(duration: 0)) { angle = 0 }
            }
        }
        .onAppear {
            if spinning && !reduceMotion {
                withAnimation(.linear(duration: 0.9).repeatForever(autoreverses: false)) {
                    angle = 360
                }
            }
        }
    }
}

// MARK: - QuickFilterPill

/// Quick-filter pill (All / Favorites / Read later) with an optional count badge. Active
/// state tints + outlines the pill and bounces it via the non-hit-testing `.bounceScale`.
struct QuickFilterPill: View {
    let label: String
    let systemImage: String
    let count: Int?
    let active: Bool
    let color: Color
    let onTap: () -> Void

    @Environment(\.curioColors) private var colors

    var body: some View {
        Button(action: onTap) {
            HStack(spacing: 5) {
                Image(systemName: systemImage)
                    .font(.system(size: 14, weight: .semibold))
                    .foregroundStyle(active ? color : colors.onSurface.opacity(0.6))
                Text(label)
                    .font(.system(size: 11, weight: .bold))
                    .tracking(0.5)
                    .foregroundStyle(active ? color : colors.onSurface.opacity(0.75))
                if let count, count > 0 {
                    Text("\(count)")
                        .font(.system(size: 10, weight: .black))
                        .foregroundStyle(active ? Color.white : colors.onSurface.opacity(0.7))
                        .padding(.horizontal, 6)
                        .padding(.vertical, 1)
                        .background(active ? color : colors.onSurface.opacity(0.12), in: Capsule())
                }
            }
            .padding(.horizontal, 12)
            .padding(.vertical, 7)
            .background(
                (active ? color.opacity(0.18) : colors.onSurface.opacity(0.05)),
                in: Capsule()
            )
            .overlay { Capsule().stroke(active ? color.opacity(0.55) : Color.clear, lineWidth: 1) }
        }
        .buttonStyle(.curioPressBounce(pressedScale: 0.93))
        .bounceScale(active: active)
    }
}

// MARK: - TypingDots

/// Three staggered, pulsing dots — the chat "assistant is typing" indicator. Each dot scales
/// 0.5→1.0 on a 600ms reversing cycle, staggered by 150ms, with alpha tracking the scale.
struct TypingDots: View {
    let color: Color

    @Environment(\.accessibilityReduceMotion) private var reduceMotion
    @State private var animate = false

    var body: some View {
        HStack(spacing: 4) {
            ForEach(0..<3, id: \.self) { i in
                Dot(color: color, index: i, animate: animate, reduceMotion: reduceMotion)
            }
        }
        .onAppear { animate = true }
    }

    private struct Dot: View {
        let color: Color
        let index: Int
        let animate: Bool
        let reduceMotion: Bool

        var body: some View {
            // Static endpoint scale when Reduce Motion is on (no infinite animation).
            let scale: CGFloat = reduceMotion ? 0.75 : (animate ? 1.0 : 0.5)
            Circle()
                .fill(color.opacity(0.4 + Double(scale) * 0.5))
                .frame(width: 7, height: 7)
                .scaleEffect(scale)
                .animation(
                    reduceMotion
                        ? nil
                        : .easeInOut(duration: 0.6)
                            .repeatForever(autoreverses: true)
                            .delay(Double(index) * 0.15),
                    value: animate
                )
        }
    }
}

// MARK: - Source glyph centralizer
//
// Centralizes the source-type → SF Symbol mapping that recurs across the card avatar/cover
// (CONVENTIONS §8: centralize source-type→glyph to avoid drift). Substitutions vs the Android
// Material icons:
//   GITHUB        : Icons.Filled.Hub                 → "point.3.connected.trianglepath.dotted"
//   ARXIV         : Icons.AutoMirrored.Filled.MenuBook → "book"
//   HUGGING_FACE  : Icons.Filled.AutoAwesome         → "sparkles"
//   else / nil    : Icons.Filled.Bookmarks           → "bookmark.fill"

/// Maps a `SourceType` (or nil) to the SF Symbol glyph used for its avatar/cover watermark.
func sourceGlyph(_ sourceType: SourceType?) -> String {
    switch sourceType {
    case .GITHUB: return "point.3.connected.trianglepath.dotted"
    case .ARXIV: return "book"
    case .HUGGING_FACE: return "sparkles"
    default: return "bookmark.fill"
    }
}

// MARK: - bodySmall (lineHeight 19) helper role

extension CurioFont {
    /// `MaterialTheme.typography.bodySmall.copy(lineHeight = 19.sp)` — the markdown panel body
    /// role (bodySmall is 12pt; line height bumped to 19 to match the Compose copy).
    static let bodySmallLineHeight19 = CurioTextStyle(size: 12, weight: .regular, tracking: 0.4, lineHeight: 19)
}

// MARK: - String.isBlank parity (Kotlin)

extension String {
    /// Kotlin `String.isBlank()`: empty or whitespace-only. (Mirrors `isNullOrBlank` checks
    /// once the optional has been unwrapped at the call site.)
    var isBlank: Bool { trimmingCharacters(in: .whitespacesAndNewlines).isEmpty }
}
